### PR TITLE
feat(nats): support nats subscribe & request

### PR DESCRIPTION
**Pull request by idonava**

### CODE REVIEW

File: src/wrappers/nats.js [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+const {
+    tracer,
+    tryRequire,
+    eventInterface,
+    utils,
+} = require('epsagon');
+const traceContext = require('../trace_context.js');
+
+
+const NATS_TYPES = {
```
Comment (ranrib) [inline src/wrappers/nats.js]: Some of it was already defined in epsagon-node, so keep it in there and import it to here

File: src/wrappers/nats.js [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+    name: 'nats',
+    mainWrappedFunction: 'Client',
+    serverDefaultHostname: 'unknown',
+    inboxSignature: '_INBOX',
+};
+
+const getServerHostname = (currentServer) => {
```
Comment (ranrib) [inline src/wrappers/nats.js]: same, use from epsagon-node

File: src/wrappers/nats.js [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+    return serverHostname;
+};
+
+/**
+ * Checking if subscribe subject is inbox listener id.
+ *
+ * @param {*} subject subsribe subject.
+ * @returns {Boolean} true if is inbox listener id, else false.
+ */
+const isNatsRequestCall = (subject) => {
```
Comment (ranrib) [inline src/wrappers/nats.js]: `const isNatsRequestCall = (subject) => (subject && typeof subject === 'string' && subject.startsWith(NATS_TYPES.inboxSignature))`

Comment (idonava) [reply src/wrappers/nats.js]: fixed.

File: src/wrappers/nats.js [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+
+const NATS_TYPES = {
+    name: 'nats',
+    mainWrappedFunction: 'Client',
+    serverDefaultHostname: 'unknown',
+    inboxSignature: '_INBOX',
+};
+
+const getServerHostname = (currentServer) => {
+    let serverHostname = NATS_TYPES.serverDefaultHostname;
```
Comment (galbash) [inline src/wrappers/nats.js]: same comment, use `a ? b : c`

File: src/wrappers/nats.js [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+            'nats',
+            serverHostname,
+            isRequestCall ? 'requestMessageListener' : 'subscribeMessageListener',
+            'trigger'
+        );
+        tracer.addEvent(natsEvent);
+
+        // Getting message data.
+        const triggerMetadata = {};
+        if (callback_msg) triggerMetadata.msg = callback_msg;
```
Comment (galbash) [inline src/wrappers/nats.js]: always use curly braces :)

File: src/wrappers/nats.js [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+            originalHandlerSyncErr = err;
+        }
+        const originalHandlerName = callback.name;
+        if (originalHandlerName) {
+            nodeEvent.getResource().setName(originalHandlerName);
+        }
+        // Handle and finalize async user function.
+        if (utils.isPromise(runnerResult)) {
+            let originalHandlerAsyncError;
+            runnerResult.catch((err) => {
```
Comment (galbash) [inline src/wrappers/nats.js]: you get the err and res in the finally clause, the catch is redundent

Comment (galbash) [inline src/wrappers/nats.js]: reassign runner result, so the user will chain his `thens` after ours

File: src/wrappers/nats.js [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+        } else {
+            // Finalize sync user function.
+            eventInterface.finalizeEvent(nodeEvent, nodeStartTime, originalHandlerSyncErr);
+            tracer.sendTrace(() => {});
+        }
+        tracer.addRunner(nodeEvent, runnerResult);
+    } catch (err) {
+        tracer.addException(err);
+    }
+    // Throwing error in case of sync user function.
```
Comment (galbash) [inline src/wrappers/nats.js]: where is `return runnerResult`?

File: src/wrappers/nats.js [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+        }
+        // Handle and finalize async user function.
+        if (utils.isPromise(runnerResult)) {
+            let originalHandlerAsyncError;
+            runnerResult.catch((err) => {
+                originalHandlerAsyncError = err;
+                throw err;
+            }).finally(() => {
+                eventInterface.finalizeEvent(nodeEvent, nodeStartTime, originalHandlerAsyncError);
+                tracer.sendTrace(() => {});
```
Comment (galbash) [inline src/wrappers/nats.js]: you should clear the tracer at some point (after the trace was sent)

File: src/wrappers/nats.js (hunk 7 above)
Comment (idonava) [reply src/wrappers/nats.js]: this function doesn't return anything, but you right, its always better to return.

File: src/wrappers/nats.js (hunk 5 above)
Comment (idonava) [reply src/wrappers/nats.js]: changed.

File: src/wrappers/nats.js (hunk 6 above)
Comment (idonava) [reply src/wrappers/nats.js]: we cannot get the error in the finally clause.

File: src/wrappers/nats.js [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+    callback_subject,
+    callback_sid,
+    callback,
+    serverHostname,
+    isRequestCall
+) {
+    let originalHandlerSyncErr;
+    let runnerResult;
+    try {
+        // Initialize tracer and evnets.
```
Comment (galbash) [inline src/wrappers/nats.js]: Typo

File: src/wrappers/nats.js [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+                        callback_reply,
+                        callback_subject,
+                        callback_sid,
+                        callback_internal,
+                        serverHostname,
+                        isRequestCall
+                    )
+                );
+            };
+            clientRequest = wrappedFunction.apply(this, [subject, opts_internal, patchedCallback]);
```
Comment (galbash) [inline src/wrappers/nats.js]: If the wrapped function throws an exception you call it twice..

File: src/wrappers/nats.js [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+        return connectFunctionResponse;
+    };
+}
+
+module.exports = {
+    /**
+     * Initializes the nats tracer
+     */
+    init() {
+        const nats = tryRequire('nats');
```
Comment (galbash) [inline src/wrappers/nats.js]: Why are you tryRequiring? Use the module utils?

File: src/wrappers/nats.js (hunk 9 above)
Comment (idonava) [reply src/wrappers/nats.js]: fixed

File: src/wrappers/nats.js (hunk 10 above)
Comment (idonava) [reply src/wrappers/nats.js]: fixed.

File: src/wrappers/nats.js (hunk 11 above)
Comment (idonava) [reply src/wrappers/nats.js]: we don't export module utils.

File: src/wrappers/nats.js [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+/**
+ * Wraps nats connect function.
+ * @param {Function} connectFunction nats connect function.
+ * @return {Function} nats connect function response.
+ */
+function natsConnectWrapper(connectFunction) {
+    traceContext.init();
+    tracer.getTrace = traceContext.get;
+    return function internalNatsConnectWrapper(url, opts) {
+        const connectFunctionResponse = connectFunction(url, opts);
```
Comment (galbash) [inline src/wrappers/nats.js]: shouldn't you use apply?

File: src/wrappers/nats.js [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+        if (connectFunctionResponse && connectFunctionResponse.constructor) {
```
Comment (galbash) [inline src/wrappers/nats.js]: no try / catch?

File: src/wrappers/nats.js (hunk 12 above)
Comment (idonava) [reply src/wrappers/nats.js]: changed.

File: src/wrappers/nats.js (hunk 13 above)
Comment (idonava) [reply src/wrappers/nats.js]: added.

**Pull request merged**